### PR TITLE
SCC-3600 - Search Results API route

### PR DESCRIPTION
**Pull request by dgcohen**
This PR introduces the following:

1. The default route handler function for Search Results
2. The fetchResults function, invoked by the handler and to be called directly from Components' getServerSideProps method in Next for Server-Side Rendering.
3. A utility function that builds a query string from a Search Parameters object, as well as some supporting functions to make it more legible.
4. A utility function that converts the query object from a HTTP request into a Search Parameters object, which can then be used to build out the appropriate query string for the request.

Additional miscellaneous changes
1. I've added typescript type checking to the pre-commit hooks to catch errors.
2. Prettier formatting on the recently merged travis config file
3. Overall app structure changes, many files moved into the /src directory, types and utils broken up into categories, etc.

Please note that DRBB and Search Results filters are not part of this work.

### CODE REVIEW

**Comment by charmingduchess** [comment]
@danamansana @EdwinGuzman can we do another joint code review on this tomorrow?

**Review by EdwinGuzman — COMMENTED** [review]
Some comments.

File: pages/api/search.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
+import type {
+  SearchParams,
+  SearchResultsResponse,
+} from "../../src/types/searchTypes"
+import type { NextApiRequest, NextApiResponse } from "next"
```
[the remark this answers is not in the document]
Comment (dgcohen) [reply pages/api/search.ts]: Good call! done

File: pages/api/search.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        : searchKeywords,
+  })
+
+  const aggregationQuery = `/aggregations?${queryString}`
+  const resultsQuery = `?${queryString}&per_page=${RESULTS_PER_PAGE.toString()}`
+
+  // Get the following in parallel:
+  //  - search results
+  //  - aggregations
+  Promise.all([
```
[the remark this answers is not in the document]
Comment (dgcohen) [reply pages/api/search.ts:55]: It appears there is a performance advantage to Promise.all in the case of errors. From what I can tell, Promise.all has a "fail fast" behavior that triggers an error in the case of one of the promises failing, whereas multiple await requires all of the calls to finish executing.

File: src/types/searchTypes.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  per_page?: string
+  q?: string
+  sort?: string
+  sort_direction?: string
+  sort_scope?: string
+  search_scope?: string
+  filters?: SearchFilters
+}
+
+export interface SearchResultsResponse {
```
[the remark this answers is not in the document]
Comment (dgcohen) [reply src/types/searchTypes.ts:40]: This type is still being used after the removal of the unused var to define the responses from the API

File: src/utils/searchUtils.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+/**
+ * getFieldQuery
+ * Get the search param from the field selected.
+ */
+function getFieldQuery(field: string = ""): string {
+  if (!field || field.trim() === "all") {
+    return ""
+  }
+  return `&search_scope=${field}`
```
[the remark this answers is not in the document]
Comment (dgcohen) [reply src/utils/searchUtils.ts:38]: This logic currently isn't being hit when a user types into the search bar, but will be as part of the upcoming PR that implements in the UI. This behavior is caused by the search_scope being set as a query param in the Search component, but will eventually be sanitized through the pre-processing utils.

File: pages/api/search.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (EdwinGuzman) [reply pages/api/search.ts:58]: Yes, that's true. I wasn't pointing out that we shouldn't use Promise.all, more so where the await should go.
```
Promise.all([
  await ....,
  await ...
])
```
vs
```
await Promise.all([
  ....,
  ....
])
```

File: package.json [hunk 5]
```diff
@@ -20,10 +20,12 @@
     "next": "12.3.4",
     "react": "17.0.2",
     "react-dom": "17.0.2",
-    "save-dev": "^0.0.1-security"
+    "save-dev": "^0.0.1-security",
```
Comment (EdwinGuzman) [inline package.json:23]: What's `save-dev` used for?

**Review by EdwinGuzman — APPROVED** [review]
This looks okay to start. The filter-related code could've been a new PR or commented out because it's not all used (such as `getFilterQuery`). Otherwise, it's a good first step. I'm sure we'll have to make modifications once filters and DRB searching is added.

File: package.json (hunk 5 above)
Comment (dgcohen) [reply package.json:23]: Good catch, this was added accidentally, must have had a typo while using the --save-dev flag.

File: src/types/searchTypes.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  dateAfter?: string
+  dateBefore?: string
+  issuance?: string[]
+}
+
+export interface IdentifierNumbers {
+  issn?: string
+  isbn?: string
+  oclc?: string
+  lccn?: string
```
[the remark this answers is not in the document]
Comment (charmingduchess) [reply src/types/searchTypes.ts:20]: Not sure how to describe them beyond identifier numbers! They are identifiers assigned by different institutions (library of congress, publishers, ohio college library center, et al) used to... identify different books. Theoretically supposed to be unique, but in practice often not.

**Pull request merged**